### PR TITLE
Change to webClient

### PR DESCRIPTION
**Pull request by Cactu5**
Changes from using the default `HttpsURLConnection` to using Spring's `webClient`.

### CODE REVIEW

File: decentralizediso20022/src/main/resources/application-external.properties [hunk 1]
```diff
@@ -2,13 +2,13 @@ server.port=443
...
 # dummy keystore
-server.ssl.key-store=src/main/resources/keystore.p12
+server.ssl.key-store=keystore.p12
 server.ssl.key-store-password=qwerty
 
 security.require-ssl=true
 
 # dummy truststore
-server.ssl.trust-store=src/main/resources/truststore.p12
+server.ssl.trust-store=truststore.p12
```
Comment (Mechakhan) [inline decentralizediso20022/src/main/resources/application-external.properties:11]: Tomcat is responsible for TLS server support on the external component. server.ssl.key-store and server.ssl.trust-store are read by tomcat, and tomcat does not know where the resources folder is. That is why I prefixed the keystore and truststore with the path to the resources folder here, but not in the internal component.

Comment (Mechakhan) [reply decentralizediso20022/src/main/resources/application-external.properties:11]: Because otherwise tomcat cannot find the files

Comment (Cactu5) [reply decentralizediso20022/src/main/resources/application-external.properties]: I have fixed this, but it did not solve the underlying issue.

**Comment by Mechakhan** [comment]
Is the use of WebClient something that should be applied to `IntMessageService` too?

**Comment by Cactu5** [comment]
> Is the use of WebClient something that should be applied to `IntMessageService` too?

You should use the `webClient` when making requests, but it isn't required for the code to work.

**Comment by Mechakhan** [comment]
> > Is the use of WebClient something that should be applied to `IntMessageService` too?
> 
> You should use the `webClient` when making requests, but it isn't required for the code to work.

Should we open an issue for it then? (Perhaps in our local repo)

**Comment by Cactu5** [comment]
> > > Is the use of WebClient something that should be applied to `IntMessageService` too?
> > 
> > 
> > You should use the `webClient` when making requests, but it isn't required for the code to work.
> 
> Should we open an issue for it then? (Perhaps in our local repo)

Sure

File: decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/presentation/external/ExtPacsController.java [hunk 2]
```diff
@@ -62,16 +55,18 @@ public ResponseEntity handlePacs008(@RequestBody String pacs) throws Unsupported
...
-        } catch (Exception e) {
-            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "The entity had an invalid from or to header.");
+            vs.validateHeader(header, port);
+        } catch (Throwable e) {
+            // Note: Might want check for WebClientResponseException, instead of only
+            // NotFound.
+            if (!(e instanceof WebClientResponseException.NotFound)) {
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
+                        "The entity had an invalid from or to header.");
+            }
```
Comment (Mechakhan) [inline decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/presentation/external/ExtPacsController.java:69]: If I've understood correctly, if `e` is not an instance of `WebClientResponseException.NotFound`, the ISO 20022 message will be forwarded to the bank system even though the header was invalid. Is there any reason why a `ResponseStatusException` is not thrown regardless of how `vs.validateHeader` failed?

File: decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/util/HTTPSCustomTruststore.java [hunk 3]
```diff
@@ -19,18 +24,51 @@
...
+     * @throws InvalidAlgorithmParameterException
      */
+    public static TrustManagerFactory createTrustManager(String truststorePath, String password)
+            throws KeyStoreException, NoSuchAlgorithmException, CertificateException, IOException,
+            InvalidAlgorithmParameterException {
+        if (truststorePath == null || password == null) {
+            throw new KeyStoreException("Truststore and password cannot be NULL.");
+        }
+
+        // KeyStore keystore = KeyStore.getInstance("PKCS12");
```
Comment (Mechakhan) [inline decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/util/HTTPSCustomTruststore.java:47]: This isn't needed anymore right?

Comment (Cactu5) [reply decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/util/HTTPSCustomTruststore.java:47]: Yes, I'll remove it.

File: decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/presentation/external/ExtPacsController.java (hunk 2 above)
Comment (Cactu5) [reply decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/presentation/external/ExtPacsController.java:69]: I need to do some further testing, but if the header is invalid it should not be forwarded. I'm only catching the `WebClientResponseException.NotFound`, no exceptions related to SSL.

The reason I had to do this is that on `localhost` I make a request to the domain's `/` endpoint, which is not configured to anything. Maybe we should not accept the certificate even if it is valid if the webpage returns a 404, or anything else than a 200?

Comment (Mechakhan) [reply decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/presentation/external/ExtPacsController.java:69]: The only thing we want to confirm is that the hostname has a valid certificate, even if requesting their `/` results in a 404 or another error code. We shouldn't require the banks to have a valid webpage on their `/` right?

Comment (Cactu5) [reply decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/presentation/external/ExtPacsController.java:69]: That's why I don't throw an error on 404.

Comment (Cactu5) [reply decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/presentation/external/ExtPacsController.java:69]: An why I think we might not want to throw an error on any `WebClientResponseException`.

Comment (Mechakhan) [reply decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/presentation/external/ExtPacsController.java:69]: OK,  I'm with you that we should ignore `WebClientResponseException`s. Coudl you change it to that and add a TODO that we should check if there are any other exceptions that we should ignore?

Comment (Cactu5) [reply decentralizediso20022/src/main/java/com/centiglobe/decentralizediso20022/presentation/external/ExtPacsController.java:69]: I'll do that

**Review by Mechakhan — APPROVED** [review]
LGTM!

**Pull request merged**